pixels-turbo/pixels-worker-vhive/src/main/java/io/pixelsdb/pixels/worker/vhive/PixelsWriterStreamImpl.java

```diff
@@ -0,0 +1,830 @@
+            logger.error(e.getMessage());
+            throw e;
+        }
+
+        // update index and stats
+        rowGroupDataLength = 0;
+        for (int i = 0; i < columnWriters.length; i++) {
+            ColumnWriter writer = columnWriters[i];
+            rowGroupDataLength += writer.getColumnChunkSize();
+            if (CHUNK_ALIGNMENT != 0 && rowGroupDataLength % CHUNK_ALIGNMENT != 0) {
+                /*
+                 * Issue #519:
+                 * This line must be consistent with how column chunks are padded above.
+                 * If we only pad after each column chunk when writing it into the physical writer
+                 * without checking the alignment of the current position of the physical writer,
+                 * then we should not consider curRowGroupOffset when calculating the alignment here.
+                 */
+                rowGroupDataLength += CHUNK_ALIGNMENT - rowGroupDataLength % CHUNK_ALIGNMENT;
+            }
+            /* TODO: writer.reset() does not work for partitioned file writing, fix it later.
+             * The possible reason is that: when the file is partitioned, the last stride of a row group
+             * (a.k.a., partition) is likely not full (length < pixelsStride), thus if the writer is not
+             * reset correctly, the strides of the next row group will not be written correctly.
+             * We temporarily fix this problem by creating a new column writer for each row group.
+             */
+            // writer.reset();  // This seems to be slower than creating a new writer for each row group.
+            columnWriters[i] = newColumnWriter(children.get(i), columnWriterOption);
+        }
+
+        if (rowGroupDataLength + 32 < recordedRowGroupDataLen || rowGroupDataLength + 8 > recordedRowGroupDataLen)  // XXX: if (rowGroupDataLength != recordedRowGroupDataLen)
+            throw new IOException("The calculated rowGroupDataLength is not equal to the recorded value");
+
+        if (partitioned) {
+            // partitionColumnIds has been checked to be present in the builder.
+            curPartitionInfo.addAllColumnIds(partKeyColumnIds.orElse(null));
+            curPartitionInfo.setHashValue(currHashValue);
+        }
+
+        // put curRowGroupIndex into rowGroupFooter
+        StreamProto.StreamRowGroupFooter.Builder rowGroupFooterBuilder =
+                StreamProto.StreamRowGroupFooter.newBuilder()
+                        .setRowGroupIndexEntry(curRowGroupIndex.build())
+                        .setRowGroupEncoding(curRowGroupEncoding.build())
+                        .setNumberOfRows(curRowGroupNumOfRows);
+        if (partitioned) {
+            rowGroupFooterBuilder.setPartitionInfo(curPartitionInfo.build());
+        }
+        StreamProto.StreamRowGroupFooter rowGroupFooter = rowGroupFooterBuilder.build();
+        // XXX: rowGroupIndex and rowGroupEncoding are the same for all row groups in the same file?
```

Comment:
rowGroupIndex: no, it contains the offsets of the column chunks.
rowGroupEncoding: yes
Please revise this comment.